### ISSUE TITLE
Auth: OpenID authentication + API for authorization

### ISSUE THREAD

**Issue by EduardHL** [metabase/metabase#3101]
Hi, congratulations for building such a good product.

Would it be possible to authenticate users with any OpenID Connect provider? The PR #2818 is specific to Google. It would be great if we could have something more generic.

To provide more context: we would like to centralise authentication and authorisation in one place for the different front-ends and APIs we provide to our enterprise users. In order to achieve this we are looking at options like [OpenAM](https://forgerock.org/openam/) or [CAS](https://github.com/apereo/cas). The advantage of having a centralised authentication and authorisation server is that you don't need to duplicate roles and permissions in the different front-ends and APIs. Our enterprise users would ultimately be authenticated with LDAP; using an authentication and authorisation server would allow us to become an OpenID connect provider and check credentials with LDAP. Should #1488 advance further, direct LDAP integration from Metabase could also be an option.

Concerning authorization and related to #3088 and #1175, it would be very interesting for us if there could be a generic API/endpoint that we can implement to check if a user can access a specific resource, in the same way in OAuth 2.0 a resource server (server hosting the protected resources; in this case Metabase) may interact with the authorization server (server issuing access tokens; our central authentication and authorization server) to check if a user has access to a resource; this would avoid defining roles for users and resources directly in Metabase.

Let us know if these widely adopted protocols and solutions (OpenID connect, separating authorisation server and resource server concerns) would make sense for Metabase and how we can help to make it happen.

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

**Comment by salsakran** [comment]
regarding " it would be very interesting for us if there could be a generic API/endpoint that we can implement to check if a user can access a specific resource", are you referring to an API you'd program against in a fork? E.g. where for each card that function would call another server and ask if user X could access that item?

**Comment by EduardHL** [comment]
Yes, exactly. It's good practice in software development to define interfaces (to decouple implementation from how functionality is used throughout the application); it makes your app more modular and easier to refactor sections of the code (which is often necessary sooner than one would desire). In this case it would suffice to have an interface with a method `canAccess` with two arguments: `resource` and `user`. The `resource` could be a datasource, card or even a dashboard (although it probably makes more sense to define permissions at the datasource/column level and figure out from there if a user has access to a card or dashboard). IMO defining interfaces has very little overhead and many advantages; for example, it would allow other teams contribute plugins (or fork), while you can focus on your target users and let others implement LDAP, other type of visualisations, etc.

**Comment by selfscrum** [comment]
8 years later... :) I have a big enterprise client waiting for this enablement. Is there some projection already if/when you start working on this feature?

**Comment by paoliniluis** [comment]
@selfscrum please write to sales so we can weight in the fact that there’s a big company that needs this

**Comment by tjfroese** [comment]
This Feature will be great.

**Comment by akzincsystems** [comment]
Your current authentication/authorization methods limit the use of the product in the AWS space using Cognito. Supporting OIDC would enable us to at least hack a solution. Supporting Cognito would be even better.

**Comment by paoliniluis** [comment]
@akzincsystems as I mentioned in my previous comment on the other issue you wrote, check https://github.com/zbodi74/metabase-cognito-auth/tree/master

**Comment by yinzara** [comment]
> [@selfscrum](https://github.com/selfscrum) please write to sales so we can weight in the fact that there’s a big company that needs this

Ditto. We're evaluating using Metabase long term but this issue being fixed is necessary or we can't purchase the Enterprise edition.
